torch/distributed/tensor/experimental/_attention.py

```diff
@@ -1086,6 +1115,74 @@ def backward_hook(grad: torch.Tensor) -> None:
         return tuple(out)
 
 
+def rewrite_context_parallel_block_mask(
+    block_mask: BlockMask,
+    shard_q_size: int,
+    cp_rank: int,
+    cp_group_size: int,
+    device_type: str,
+) -> BlockMask:
+    def _rewrite_mask_mod(
+        mask_mod: _mask_mod_signature,
+        rank: int,
+        world_size: int,
+        block_size: int,
+        local_q_size: int,
+    ) -> _mask_mod_signature:
+        def local_q_idx_to_q_idx(local_q_idx: torch.Tensor) -> torch.Tensor:
+            # calculate local block_idx and block_offset
+            local_blk_idx, local_blk_offset = (
+                local_q_idx // block_size,
+                local_q_idx % block_size,
+            )
+            # NOTE: load balancing is not used
+            local_num_blocks = local_q_size // block_size
+            blk_idx = local_num_blocks * rank + local_blk_idx
+            return blk_idx * block_size + local_blk_offset
+
+        return lambda b, h, q_idx, kv_idx: mask_mod(
+            b,
+            h,
+            local_q_idx_to_q_idx(q_idx),
+            kv_idx,
+        )
+
+    def _create_block_mask(
```

Comment:
Do we need this?

Reply:
good catch! I added this because I was planning to add `lru_cache` but actually this cache is not needed because I'm now requiring user to explicitly initialize the `BlockMask`s instead of implicitly `create_block_mask` in `FlexAttentionHOP` dispatch.

Reply:
I typically create a global `create_block_mask_compiled` and use that one